tools/insn_queue.c

```diff
@@ -0,0 +1,251 @@
+/*
+ *  libpulp - User-space Livepatching Library
+ *
+ *  Copyright (C) 2023 SUSE Software Solutions GmbH
+ *
+ *  This file is part of libpulp.
+ *
+ *  libpulp is free software; you can redistribute it and/or
+ *  modify it under the terms of the GNU Lesser General Public
+ *  License as published by the Free Software Foundation; either
+ *  version 2.1 of the License, or (at your option) any later version.
+ *
+ *  libpulp is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ *  Lesser General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with libpulp.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "insn_queue.h"
+#include "introspection.h"
+#include "ptrace.h"
+
+#include <assert.h>
+#include <stdio.h>
+
+/** PID of target process in which we will execute the instructions.  */
+static int remote_pid;
+
+/** @brief Interpret a print instruction.
+ *
+ * @param insn    Instruction to interpet. Must be a print instruction.
+ *
+ * @return        Size of interpreted instruction.
+ */
+static int
+insn_interpret_print(struct ulp_insn *insn)
+{
+  struct ulp_insn_print *p_insn = (struct ulp_insn_print *)insn;
+
+  printf("%s\n", p_insn->bytes);
+  return insn->size;
+}
+
+/** @brief Interpret NOP instruction.
+ *
+ * @param insn    Instruction to interpet. Must be a NOP instruction.
+ *
+ * @return        Size of interpreted instruction (always 1 byte).
+ */
+static int
+insn_interpret_nop(struct ulp_insn *insn)
+{
+  (void)insn;
+  return 1;
```

Comment:
I think your instructions are always at least 4 bytes, as the insn header is that larger (it always includes the size member).  So returning '1' here is wrong.  As you don't have routines in the libpulp side to actually generate such NOP instruction you won't see a problem, but still this should be fixed.

Reply:
Actually the NOP insn was supposed to take only one byte, but since now I am careful about alignment issues it is sensible to expand this to 4.